File: java/src/CardGame/CardGame.java
package CardGame;
import Display.ConsoleOutput;
import Display.ConsoleInput;
import Display.Input;
import Display.Output;
import Player.Player;
import Player.ComputerPlayer;
import Player.PlayerType;
import BlackJackWithState.BlackJackComputerPlayer;

import java.util.ArrayList;

public class CardGame {

    public Integer noOfCards = 7;
    private Deck deck;
    private ArrayList<Player> players;
    private Input userInput;
    private Output userOutput;
    private CardGame game;

    public Input getUserInput() {
        return userInput;
    }

    public Output getUserOutput(){
        return this.userOutput;
    }

    public CardGame getGame(){
        return game;
    }

    public Deck getDeck() { return deck;}

    public ArrayList<Player> getPlayers(){
        return players;
    }

    protected void setGame(CardGame game){
        this.game = game;
    }

    public CardGame(){
        this.deck = Deck.getInstance();
        this.userInput = new ConsoleInput();
        this.userOutput = new ConsoleOutput();
        this.players = new ArrayList<Player>();
    }

    private void createHumanPlayer(){
        userOutput.output("What is your name");
        String name = userInput.getInputString();
        players.add(new Player(PlayerType.USER,name));
    }

    private void createComputerPlayers(Integer noOfPlayers) {
        ComputerPlayer dealer = new ComputerPlayer(PlayerType.DEALER,"Dealer",17);
        players.add(dealer);
        players.add(new ComputerPlayer(PlayerType.COMPUTER,"Comp 1",0));
        noOfPlayers -= 1; //Remove the dealer
        for (int counter=2;counter < noOfPlayers;counter++){
            players.add(new ComputerPlayer(PlayerType.COMPUTER,"Comp" + counter,0));
        }
    }

    protected void initiatePlayers(){
        createHumanPlayer();
        userOutput.output("How many players, minimum of two?");
        int noOfPlayers = userInput.getInputInt();
        createComputerPlayers(noOfPlayers);
        dealCards(noOfPlayers);

    }

    public void dealCards(Integer noOfPlayers){
        boolean allCards = false;
        if (noOfCards == 0) {
            noOfCards = (int) Math.floor(deck.getNumberOfCards()/noOfPlayers);
            allCards = true;
        }
        for (Player player: players){
            player.setHand(dealHand(deck.getDeck(),noOfCards));
        }
        if (allCards){
            for (Player player: players){
                if (deck.getNumberOfCards() > 0){
                    player.getHand().add(deck.playACard());
                }
            }
        }
    }

    public Hand dealHand(Hand hand, int noOfCards){

        for (int cardCounter=0;cardCounter<noOfCards;cardCounter++){
            if (deck.getNumberOfCards() > 0) {
                hand.add(deck.playACard());
            }
        }
        return hand;
    }


    public void initiate(){
        initiatePlayers();
    }

    public void play(){
        initiate();

    }

    public int getScore(Hand hand){
        return 0;
    }

    public void showPlayers(){
        for (Player player: players){
            userOutput.output(player.getName());
            userOutput.output(player.getHand().show());

        }
    }

    public static void main(String[ ] args) {
        CardGame cardGame = new CardGame();
        cardGame.play();
        cardGame.showPlayers();


    }
}

File: python/src/CardGame/CardGame.py
from python.src.CardGame.Deck import Deck
from python.src.CardGame.Hand import Hand
import math
class CardGame():

    noOfPlayers = None
    noOfCards = 0
    hands = []
    deck = Deck()

        # Function: dealCards
    # Description: Deal a number of cards from a deck to a number of players (no of hands). Optionally an existing set of
    # hands of cards can be passed. The default is for no existing hands of cards to be passed.
    def dealCards(self,noOfCards,noOfHands):
        hands=[]
        allCards = False
        if noOfCards == 0:
            noOfCards = math.floor(self.deck.size()/noOfHands)
            allCards = True

        for handIndex in range(0,noOfHands):
            hand = Hand()
            for counter in range(noOfCards):
                hand.append(self.deck.dealACard())
            hands.append(hand.copy())

        if allCards:
            counter = 0
            while self.deck.size() > 0:
                hands[counter].append(self.deck.dealACard())
                counter = (counter +1) % noOfHands

        return hands

    def startCardGame(self):
        self.noOfPlayers = int(input("Enter the number of players"))
        self.deck.generateDeck()
        self.deck.shuffleCards()
        self.hands = self.dealCards(self.noOfCards,self.noOfPlayers)

if __name__ == '__main__':
    cardGame = CardGame()
    cardGame.startCardGame()
    for hand in cardGame.hands:
        print(hand.getHand())
        print(hand.size())
    print(cardGame.deck.size())
